lapack/testlapack/dgesc2.go

```diff
@@ -12,80 +12,92 @@ import (
-	// Dgesc2 solves a system of linear equations
-	//  A * X = scale * RHS
-	// with a general n×n matrix A using the LU factorization with
-	// complete pivoting computed by Dgetc2. The result is placed in
-	// rhs on exit.
 	Dgesc2(n int, a []float64, lda int, rhs []float64, ipiv, jpiv []int) (scale float64)
+
+	Dgetc2er
 }
 
 func Dgesc2Test(t *testing.T, impl Dgesc2er) {
-	const tol = 1e-12
 	rnd := rand.New(rand.NewSource(1))
-	for _, test := range []struct {
-		n, lda int
-	}{
-		{3, 0},
-		{5, 0},
-		{20, 30},
-		{200, 0},
-	} {
-		testSolveDgesc2(t, impl, rnd, test.n, test.lda, tol)
+	for _, n := range []int{0, 1, 2, 3, 4, 5, 10, 20, 50} {
+		for _, lda := range []int{n, n + 3} {
+			testDgesc2(t, impl, rnd, n, lda, false)
+			testDgesc2(t, impl, rnd, n, lda, true)
+		}
 	}
 }
 
-func testSolveDgesc2(t *testing.T, impl Dgesc2er, rnd *rand.Rand, n, lda int, tol float64) {
-	name := fmt.Sprintf("n=%v,lda=%v", n, lda)
-	if lda == 0 {
-		lda = n
-	}
+func testDgesc2(t *testing.T, impl Dgesc2er, rnd *rand.Rand, n, lda int, big bool) {
+	const tol = 1e-14
+
+	name := fmt.Sprintf("n=%v,lda=%v,big=%v", n, lda, big)
+
 	// Generate random general matrix.
-	a := randomGeneral(n, n, lda, rnd)
-	// anorm := floats.Norm(a.Data, 1)
-
-	// Generate a random solution.
-	xWant := randomGeneral(n, 1, 1, rnd)
-	// xnorm := floats.Norm(xWant.Data, 1)
-
-	// Compute RHS vector that solves for X such that  A*X = scale * RHS
-	rhs := zeros(n, 1, 1)
-	blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, a, xWant, 1, rhs)
-	rhsCopy := zeros(n, 1, 1) // Will contain A*x result.
-	copyGeneral(rhsCopy, rhs)
-	// Compute LU factorization with full pivoting.
-	lu := zeros(n, n, lda)
-	copyGeneral(lu, a)
+	a := randomGeneral(n, n, max(1, lda), rnd)
+
+	// Generate a random right hand side vector.
+	b := randomGeneral(n, 1, 1, rnd)
+	if big {
+		for i := 0; i < n; i++ {
+			b.Data[i] *= bignum
+		}
+	}
+
+	// Compute the LU factorization of A with full pivoting.
+	lu := cloneGeneral(a)
 	ipiv := make([]int, n)
 	jpiv := make([]int, n)
 	impl.Dgetc2(n, lu.Data, lu.Stride, ipiv, jpiv)
 
-	// Solve using lu factorization.
-	scale := impl.Dgesc2(lu.Rows, lu.Data, lu.Stride, rhs.Data, ipiv, jpiv)
-	x := rhs
-	if scale < 0 || scale > 1 {
-		t.Errorf("%v: resulting scale out of bounds [0,1]", name)
+	// Make copies of const input to Dgesc2.
+	luCopy := cloneGeneral(lu)
+	ipivCopy := make([]int, len(ipiv))
+	copy(ipivCopy, ipiv)
+	jpivCopy := make([]int, len(jpiv))
+	copy(jpivCopy, jpiv)
+
+	// Call Dgesc2 to solve A*x = scale*b.
+	x := cloneGeneral(b)
+	scale := impl.Dgesc2(n, lu.Data, lu.Stride, x.Data, ipiv, jpiv)
+
+	if n == 0 {
+		return
+	}
+
+	// Check that const input to Dgesc2 hasn't been modified.
+	if !floats.Same(lu.Data, luCopy.Data) {
+		t.Errorf("%v: unexpected modification in lu", name)
+	}
+	if !intsEqual(ipiv, ipivCopy) {
+		t.Errorf("%v: unexpected modification in ipiv", name)
+	}
+	if !intsEqual(jpiv, jpivCopy) {
+		t.Errorf("%v: unexpected modification in ipiv", name)
```

Comment:
"jpiv"

Reply:
Thanks, fix submitted in https://github.com/gonum/gonum/pull/1675.